### ISSUE TITLE
Using deep learning to read your thoughtswith Keras and an EEG sensor

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#7799]
Using deep learning to &ldquo;read your thoughts&rdquo;&#8202;&mdash;&#8202;with Keras and an EEG sensor<br>
<br>
https://ift.tt/2Thtden<br>
<br>
Justin Alvey<br>
<br>
<div>
<div>
<div>
<div>
<p>A typical technique such as this involves:<br>
a) converting to the frequency domain then<br>
b) filtering and weighting based on some prior understanding of what features are important</p>
<p>But from math, we also know that <strong>filtering in the frequency domain</strong> is equivalent to <strong>convolution in the time domain</strong>.</p>
<p>What I&rsquo;m leading to is that with the stacked convolutional layers of a CNN, we can a) perform the same feature recognition options directly from the time series data without having to convert to the frequency domain, and b) have the network learn filters that are able to best identify these features itself.</p>
<p>This does still require careful network architecture design and sufficient training data or even pre-training, but by doing this, we can take away the manual process of finding out the most informative spectra and waveforms, letting the neural network learn and optimize these filters itself. Seeing as we&rsquo;re not exactly sure what optimal features we should be looking for, this is a sound place to start.</p>
<h1>Creating the Convolutional Neural Network</h1>
<p>For our CNN, we will borrow from the structure of those used for image classification, yet instead of using <strong>two</strong> spatial dimensions with a depth of 3 (for each color), we will have <strong>one</strong> dimension (time) with a depth of 4 (for each EMG channel).</p>
<p>With a CNN, each successive convolutional layer will develop filters that will be able to recognize successively more complex features in the EMG data. Another benefit of CNN&rsquo;s for this application is shift invariance, thanks mostly due to the max pooling in-between convolutional layers, which will allow relevant features to be detected irrespective of their placement in time, alignment or speed at which they are &ldquo;said&rdquo;.</p>
<p>Knowing the firing rate of motor neurons, we can guess the frequencies at which the most helpful features are likely to be present, we can make some educated guesses at a network structure to detect frequencies and later phonemes.</p>
<figure>
<figcaption>Example of a 1D CNN model used</figcaption>
</figure>
<p>Firstly, we can downsample the data from 1600Hz by a factor of 4 as it is unlikely there will be helpful features we will miss, and training at this resolution will likely take up excessive compute time and overfit.</p>
<p>The first layer can have filters to cover around 25ms (i.e. kernel size of 5&ndash;10), using a balance of stride and max-pooling to reduce the width for the next layer, equivalent to taking roughly 15ms steps. The number of filters was chosen to be around 50, which should be enough to represent all the different frequencies.</p>
<p>We can then experiment with a couple more convolutional layers, reducing the width through decreasing stride and max-pooling, before global max pooling, and feeding the extracted features into two fully-connected layers, to perform the classification.</p>
<p>As we have limited data and no pre-training, dropout is a good practice to avoiding overfitting and ensure lower layers learn to recognize features.</p>
<p>We can track how the training and validation accuracy develops to get an intuition as to whether the network is complex enough to classify the data yet without overfitting.</p>
<p>To start, we can load some sessions and then shuffle and split this into our training and test data. Within a few quick runs, we can already start to settle on some hyperparameters that provide some good enough results.</p>
<figure>
<figcaption>Training (- - -) and Validation (&mdash;) accuracy for a few different network shapes <em>(Not shown &mdash; experiments with varying dropout, downsampling, number of convolutional filters, learning rates, epoch sizes&hellip;)</em></figcaption>
</figure>
<p>To validate further, we can train the model on a selected few of the sessions, and then use the model to predict the labels of a separate, unseen session, recorded with different electrodes. To compare the actual and predicted labels, we can plot a confusion matrix, showing some decent results, proving the classifier is working well on unseen, session-independent data:</p>
<p>The most &ldquo;confusion&rdquo; here was between the two syllable words. Lastly, I had recorded a session that I had forgotten the ordering of the four words I used. Using the model to predict the probabilities for each class, we can plot a visualization and pretty quickly figure out the ordering&hellip;</p>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via Medium<br>
<br>
April 16, 2021 at 04:14PM
